### ISSUE TITLE
[bug] Ratelimit is not expressed as Mastodon, should be ISO8601 not Epoch

### ISSUE THREAD

**Issue by duponin** [superseriousbusiness/gotosocial#2129]
### Describe the bug with a clear and concise description of what the bug is.

I'm trying to FediFetcher's, however it ends in ratelimit and gets fatal error because ratelimit isn't expressed as Mastodon.
It expects ratelimit to be ISO8601, but GoToSocial send it as Unix Epoch.

### What's your GoToSocial Version?

0.11.0

### GoToSocial Arch

arm64

### What happened?

GoToSocial ratelimit is a Unix epoch:

```
X-RateLimit-Reset: 1692259524
```

### What you expected to happen?

Mastodon ratelimit is ISO8601:

```
X-RateLimit-Reset: 2023-08-17T08:42:22+00:00
```

### How to reproduce it?

It's fully explained in FediFetcher's issue: https://github.com/nanos/FediFetcher/issues/78

**Comment by NyaaaWhatsUpDoc** [comment]
Rate limit header in seconds is standard behaviour, if they don't support that then that's a bug on their end.

**Comment by tsmethurst** [comment]
Mm, I think it would be less fragile behavior for a client app to be able to parse either epoch time or html datetime, which is indeed based on ISO8601 (as per https://developer.mozilla.org/en-US/docs/Web/HTML/Date_and_time_formats anyway).

A little search around for x-limit-reset shows that lots of applications seem to use one or the other format for this, with no clear standard saying which it ought to be -- at least, I can't find an RFC defining this. Some even provide number of seconds until the reset occurs, so that's a third format!

In this particular case, since we're trying to keep parity with the Mastodon API, we should probably try to return the same format, so I for one would be cool with changing this.

**Comment by nanos** [comment]
Just to chime in here that I agree that FediFetcher should probably accept either, and I'll try to address that when I got a moment.

Just a minor correction: The format that Mastodon actually returns their `X-RateLimit-Reset` in is slightly different from what @duponin described (which is my fault: I had given him this incorrect info). Mastodon actually provides this without the time zone offset, but simply with `Z` instead:

```
X-RateLimit-Reset: 2023-08-17T08:42:22Z
```

**Comment by mirabilos** [comment]
@nanos the `Z` there is the timezone offset

**Comment by daenney** [comment]
X-RateLimit-Reset doesn't have an RFC defining it unfortunately, since it's an `X-` header. Based on looking around, the most prevalent recommendation seems Unix epoch timestamp. Mastodon does the date thing, and some really odd ones do things like `6m0s` which seems like it would break everyone.

The standardisation that's underway at https://www.ietf.org/archive/id/draft-ietf-httpapi-ratelimit-headers-07.html has the headers looking quite different. They use structured headers, with a "time window" in seconds instead.

For now we're going to keep it as is in GtS and it's probably best if other implementations are more permissive since the value of that header could be anything. It's also possible for people to set or override these headers in a reverse proxy, so even if we change our format to match Mastodon's it's not guaranteed to always be what a client sees.

**Comment by daenney** [comment]
@tsmethurst reopened it, since you said you probably want to change it.

**Comment by tsmethurst** [comment]
Yep yep! My reasoning is basically that there's no standard currently, and no advantage of doing it one way or the other, and it's an easy-ish fix so we might as well do it the mastodon way. If it was a big pain I'd say let's not, but it *should* be only a small pain.
